File: frontend/services/integraciones.ts
import { apiFetch } from "@/services/api"

export interface Integracion {
  tipo: string
  email_cuenta: string | null
  activo: boolean
  connected: boolean
}

export async function fetchIntegraciones(): Promise<Integracion[]> {
  return apiFetch<Integracion[]>("/api/integraciones")
}

export async function getGoogleAuthUrl(): Promise<{ auth_url: string }> {
  return apiFetch<{ auth_url: string }>("/api/integraciones/google/auth")
}

export async function saveAnthropicKey(api_key: string): Promise<Integracion> {
  return apiFetch<Integracion>("/api/integraciones/anthropic", {
    method: "POST",
    body: JSON.stringify({ api_key }),
  })
}

export async function disconnectIntegracion(tipo: string): Promise<void> {
  await apiFetch<void>(`/api/integraciones/${tipo}`, { method: "DELETE" })
}

File: backend/routers/integraciones.py
"""
Router de integraciones por usuario.
Gestiona Google OAuth y API keys (Anthropic).
"""
from typing import Optional

from fastapi import APIRouter, Request
from fastapi.responses import RedirectResponse

from config.settings import settings
from schemas.integracion import ApiKeyUpdate, IntegracionResponse
from services.integracion_service import IntegracionService
from utils.permisos import Seccion

router = APIRouter()
SECCION = Seccion.INTEGRACIONES


def _service() -> IntegracionService:
    return IntegracionService()


@router.get("", response_model=list[IntegracionResponse])
async def list_integraciones(request: Request) -> list[IntegracionResponse]:
    user_id: str = request.state.user["id"]
    return _service().get_integraciones(user_id)


@router.get("/google/auth")
async def google_auth_url(request: Request) -> dict[str, str]:
    user_id: str = request.state.user["id"]
    auth_url = _service().init_google_oauth(user_id)
    return {"auth_url": auth_url}


@router.get("/google/callback")
async def google_callback(
    state: str,
    code: Optional[str] = None,
    error: Optional[str] = None,
) -> RedirectResponse:
    if error or not code:
        return RedirectResponse(url=f"{settings.frontend_url}/configuracion?oauth=error")
    try:
        _service().handle_google_callback(user_id=state, code=code)
    except Exception:
        return RedirectResponse(url=f"{settings.frontend_url}/configuracion?oauth=error")
    return RedirectResponse(url=f"{settings.frontend_url}/configuracion?oauth=google")


@router.post("/anthropic", response_model=IntegracionResponse)
async def save_anthropic_key(
    request: Request,
    body: ApiKeyUpdate,
) -> IntegracionResponse:
    user_id: str = request.state.user["id"]
    return _service().save_anthropic_key(user_id, body.api_key)


@router.post("/zernio", response_model=IntegracionResponse)
async def save_zernio_key(request: Request, body: ApiKeyUpdate) -> IntegracionResponse:
    user_id: str = request.state.user["id"]
    return _service().save_zernio_key(user_id, body.api_key)


@router.delete("/{tipo}", status_code=204)
async def disconnect(tipo: str, request: Request) -> None:
    user_id: str = request.state.user["id"]
    _service().disconnect(user_id, tipo)
